src/libUtils/Logger.cpp

```diff
@@ -16,325 +16,264 @@
+  }
+
+ private:
+  const LogMessage& m_message;
+};
+}  // namespace
+
+namespace std {
+// Output operator to be able to use LoggerManip manipulators.
+template <std::ostream& (*FuncT)(std::ostream&, const LogMessage&)>
+inline ostream& operator<<(std::ostream& stream,
+                           const LoggerManip<FuncT>& manip) {
+  return manip(stream);
 }
+}  // namespace std
 
 namespace {
-/// helper function to get tid with better cross-platform support
-inline pid_t getCurrentPid() {
-#if defined(__linux__)
-  return syscall(SYS_gettid);
-#elif defined(__APPLE__) && defined(__MACH__)
-  uint64_t tid64;
-  pthread_threadid_np(NULL, &tid64);
-  return static_cast<pid_t>(tid64);
-#else
-#error  // not implemented in this platform
-  return 0;
-#endif
+std::ostream& logThreadId(std::ostream& stream, const LogMessage& message) {
+  stream << '[' << std::hex
+         << PAD(message._call_thread_id, Logger::TID_LEN, ' ') << std::dec
+         << ']';
+  return stream;
 }
-}  // namespace
 
-const streampos Logger::MAX_FILE_SIZE =
-    1024 * 1024 * 100;  // 100MB per log file
+std::ostream& logTimestamp(std::ostream& stream, const LogMessage& message) {
+  // The following is taken from g3log's localtime_formatted; we're changing
+  // it here to show our own time format in UTC instead of localtime.
+  // (License is completely free; see:
+  // https://github.com/KjellKod/g3log/blob/master/LICENSE)
+  auto ts = to_system_time(message._timestamp);
+  auto format_buffer =
+      internal::localtime_formatted_fractions(ts, "%y-%m-%dT%T.%f3");
+  auto time_point = std::chrono::system_clock::to_time_t(ts);
+  auto t = std::gmtime(&time_point);
+
+  stream << '[' << g3::put_time(t, format_buffer.c_str()) << ']';
+  return stream;
+}
+
+std::ostream& logCodeLocation(std::ostream& stream, const LogMessage& message) {
+  auto fileAndLine = message._file + ':' + std::to_string(message._line);
+  stream << '[' << LIMIT_RIGHT(fileAndLine, Logger::MAX_FILEANDLINE_LEN) << "]["
+         << LIMIT(message._function, Logger::MAX_FUNCNAME_LEN) << ']';
+  return stream;
+}
+
+std::ostream& logLevel(std::ostream& stream, const LogMessage& message) {
+  stream << '[' + message.level().substr(0, 4) + ']';
+  return stream;
+}
+
+std::ostream& logMessage(std::ostream& stream, const LogMessage& message) {
+  stream << message.message();
+  return stream;
+}
+
+using ThreadId = LoggerManip<&logThreadId>;
+using Timestamp = LoggerManip<&logTimestamp>;
+using CodeLocation = LoggerManip<&logCodeLocation>;
+using Message = LoggerManip<&logMessage>;
+using Level = LoggerManip<&logLevel>;
+
+std::ostream& logMessageCommon(std::ostream& stream,
+                               const LogMessage& message) {
+  stream << ThreadId(message) << Timestamp(message) << CodeLocation(message)
+         << Message(message) << std::endl;
+  return stream;
+}
+
+// FIXME: make this configurable. This is per LogRotate sink.
+const int MAX_ARCHIVED_LOG_COUNT = 15;
```

Comment:
can we make this configurable?

Reply:
+1, unlike many other things that we should make non-configurable, this one deserves it.

Reply:
This is now configurable.